### PR TITLE
Add isSupportedEnvironment virtual member

### LINKED ISSUE #2692: Update plugin definition with isEnvironmentSupported optional member
## Type of report

Task

## Provide description of the task

Some of our plugins work only on chosen environments. As an example, `easyimage` is working only on modern browsers. It would be nice to have a dedicated place where such information is placed. The best option seems to treat `isEnvironmentSupported` as an abstract, optional member of the plugin definition. In the future, we may use this function to conditionally load plugins directly from the loader, but for now, it should be enough to just document it in the plugin definition.

### PR DESCRIPTION
**Pull request by jacekbogdanski**
## What is the purpose of this pull request?

Task

Closes #2692

### CODE REVIEW

File: core/plugindefinition.js [hunk 1]
```diff
@@ -175,3 +175,28 @@
  * @since 4.2
  * @property {Boolean} hidpi
  */
+
+/**
+ * A virtual function which should be implemented if a plugin is not supported on every
```
Comment (Comandeer) [inline core/plugindefinition.js]: It's not virtual, it's very real 😛

File: core/plugindefinition.js [hunk 2]
```diff
@@ -175,3 +175,28 @@
...
+ * ```javascript
+ * CKEDITOR.plugins.add( 'sample', {
+ *		isEnvironmentSupported: function() {
+ *			// Plugin supported only on modern browsers.
+ *			return !CKEDITOR.env.ie || CKEDITOR.env.edge;
+ *		}
+ * } );
+ * ```
+ *
+ * @since 4.11.2
```
Comment (Comandeer) [inline core/plugindefinition.js]: Probably it would be better to introduce such change in next major release, 4.12.0, not in 4.11.2.

File: core/plugindefinition.js (hunk 1 above)
Comment (jacekbogdanski) [reply core/plugindefinition.js]: https://en.wikipedia.org/wiki/Virtual_function

However, taking into account that this function is not implemented by default, `abstract` may better define its purpose.

Comment (jacekbogdanski) [reply core/plugindefinition.js]: Actually, I see that the whole [pluginDefinition](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_pluginDefinition.html) is marked as `virtual`, so I will just skip modifier to not repeat it needlessly.

**Comment by engineering-this** [comment]
It would be nice to refactor all the plugins which has limited environment support. We need to ignore tests once `isSupportedEnvironment` property/method is accessible, hence it should be in listener to `pluginsLoaded`. I'd make sense to put such logic in one general helper which would be accessible by any test.
This proposal comes from https://github.com/ckeditor/ckeditor-dev/pull/2694#discussion_r256816851

**Comment by jacekbogdanski** [comment]
I'm not sure if this member should be attached into plugin namespace. In that case, we are unable to unify this function among the plugins. AFAIK we don't have any recommendation nor abstract interface how to write a plugin namespace, so we will be unable to define it in our docs easily, which can be done with `pluginDefinition`. I see that it will require some refactoring like for `tableselection` plugin but it still has to be done for other plugins like `easyimage` which uses `isUnsupportedEnvironment` in tests (this member has to be moved into the plugin itself).

Also, if I have a simple plugin like `autolink`, I will have to create special, separate plugin namespace to keep `isSupportedEnvironment` member. Writing something like:

```js
CKEDITOR.plugins.autolink = { 
   isSupportedEnvironment: fn 
};

CKEDITOR.plugins.add( autolink, {
   init: function() { // init stuff }
 } );
```

seems less coder friendly for me than just:

```js
CKEDITOR.plugins.add( autolink, {
   init: function() { // init stuff },
   isSupportedEnvironment: fn
 } );
```

Currently, our options are:
1. Define virtual method as a member of [pluginDefinition](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_pluginDefinition.html)
2. Put this function into the plugin namespace (not sure how to document such a requirement).
3. ?

**Comment by engineering-this** [comment]
I'd go with first option.

I have one suggestion. If this method is optional, then maybe it should have default value `function(){ return true; }`. Thanks to this we could update bender so it ignores every test for unsupported env based on `bender-ckeditor-plugins` declared in test file. We wouldn't have to ever write

```javascript
if ( plugin.isSupportedEnvironment ) {
    bender.ignore(); // or assert.ignore();
 }
```

**Comment by jacekbogdanski** [comment]
I was also thinking about it. Also, it fixes the issue when you have to manually check if `isSupportedEnvironment` is declared, like:
```javascript
if ( plugin.isSupportedEnvironment && plugin.isSupportedEnvironment) {
// do stuff
}
```

**Comment by f1ames** [comment]
I'm late to the party, but agree with @engineering-this on adding virtual method to `pluginDefinition`. And integrating it with bender would be nice 🤩

**Comment by jacekbogdanski** [comment]
It seems like it makes sense to pass an optional editor instance as an argument. It will complicate refactoring a bit, although some plugins require information about editor state. As an example, `docprops` plugin only works for a classic editor with `config.fullPage = true`.

**Comment by jacekbogdanski** [comment]
I refactored `tableselection`, `easyimage` and `balloontoolbar` plugins + tests. There is much more work to do thus this PR will require reviewing all our plugins in the context of a supported environment.  I would like to take it slowly because if the current implementation will be rejected, we will spend a lot of time refactoring again.

Please, take a look if you are fine with the current changes and if so, I will continue refactoring/implementing support sniffing methods.

**Comment by f1ames** [comment]
Taking over review of this PR.

**Comment by f1ames** [comment]
Rebased on the latest `major`.

File: core/editor.js [hunk 3]
```diff
@@ -585,6 +585,11 @@
 				// language has been resolved to this plugin.
 				languageCodes.push( lang );
 
+				// Plugin is supported by default (#2692).
+				plugin.isSupportedEnvironment = plugin.isSupportedEnvironment || function() {
+					return true;
+				};
```
[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply core/editor.js]: `onLoad` method may not be provided, but yeah it makes more sense to attach default method implementation inside `core/plugins.js` file.

[the remark this answers is not in the document]
Comment (f1ames) [reply core/editor.js]: > `onLoad` method may not be provided.

Yes, I meant the line above. It shouldn't depend on presence of `onLoad`.

File: plugins/balloontoolbar/plugin.js [hunk 4]
```diff
@@ -614,6 +614,10 @@
 	CKEDITOR.plugins.add( 'balloontoolbar', {
 		requires: 'balloonpanel',
 
+		isSupportedEnvironment: function() {
+			return !CKEDITOR.env.ie || CKEDITOR.env.version > 8;
```
[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply plugins/balloontoolbar/plugin.js]: It's not the same condition:

`!CKEDITOR.env.ie || CKEDITOR.env.version > 8` -  not IE or IE above 8, so Edge, IE9-11
`!CKEDITOR.env.ie || CKEDITOR.env.edge` - not IE or Edge

File: core/editor.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply core/editor.js]: Looks like the safest place is https://github.com/ckeditor/ckeditor-dev/blob/14e3bab99f1d4b968dd44e9b4e5b408e97b26758/core/plugins.js#L37 because it makes sure that the function won't be attached multiple times.

File: tests/_benderjs/ckeditor/static/tools.js [hunk 5]
```diff
@@ -59,6 +59,41 @@
...
+		 * environment. Uses {@link CKEDITOR.pluginDefinition#isSupportedEnvironment} to
+		 * verify if plugin is supported.
+		 *
+		 * Works for both manual and unit tests.
+		 *
+		 * @param pluginName pluginName Plugin name to check.
+		 * @param CKEDITOR.editor [editor] Editor instance passsed as an argument
+		 * to the {@link CKEDITOR.pluginDefinition#isSupportedEnvironment} method.
+		 */
+		ignoreUnsupportedEnvironment: function( pluginName, editor ) {
```
[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply tests/_benderjs/ckeditor/static/tools.js]: Yep, the function is not somehow glued to the single test suite editor.

File: plugins/balloontoolbar/plugin.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (f1ames) [reply plugins/balloontoolbar/plugin.js]: > `!CKEDITOR.env.ie || CKEDITOR.env.version > 8` - not IE or IE above 8, so Edge, IE9-11

Ok, so it covers all browser except IE8 (or lower) 👍 We should have something like `CKEDITOR.env.isNotIE8` 😄

**Comment by jacekbogdanski** [comment]
As we already talked F2F, there are some popular use cases where we cannot simply automate ignoring tests, like:
* using multiple declared editors by `bender.editors`
* declared editors may use a different set of plugins than information in `bender-ckeditor-plugins` tag

**Comment by f1ames** [comment]
Rebased on the latest `major`.

File: tests/plugins/balloontoolbar/context/manual/focusablebuttons.html [hunk 6]
```diff
@@ -20,10 +20,7 @@ <h2>Inline editor</h2>
 </div>
 
 <script>
-	// Ignore on mobiles due to 'mousedown' event.
-	if ( CKEDITOR.env.ie && CKEDITOR.env.version < 9 || bender.tools.env.mobile ) {
-		bender.ignore();
-	}
+	bender.tools.ignoreUnsupportedEnvironment( 'balloontoolbar' );
```
[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply tests/plugins/balloontoolbar/context/manual/focusablebuttons.html]: Sorry, refactoring issue.

File: tests/plugins/easyimage/uploadintegrations.js [hunk 7]
```diff
@@ -27,7 +27,7 @@
 		tests = {
 			init: function() {
 				// We need to ignore entire test suit to prevent of fireing init, which breaks test suit on IE8-IE10.
-				if ( easyImageTools.isUnsupportedEnvironment() ) {
+				if ( !this.editor.plugins.easyimage.isSupportedEnvironment() ) {
```
[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply tests/plugins/easyimage/uploadintegrations.js]: This condition uses `bender.ignore` to ignore the whole test suite, however, it's a unit test so `assert.ignore` is used by `ignoreUnsupportedEnvironment` function. I was thinking about improving the ignoring method but it seems not worth complicating it for the single edge case.

File: tests/plugins/image2/init.js [hunk 8]
```diff
@@ -24,6 +22,8 @@
 
 				spy.restore();
 
+				bender.tools.ignoreUnsupportedEnvironment( 'easyimage', editor );
```
[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply tests/plugins/image2/init.js]: To make sure it's really ignored. Removed :smile:

File: tests/plugins/tableselection/integrations/core/selection.js [hunk 9]
```diff
@@ -93,6 +92,10 @@
 	}
 
 	var tests =  {
+		setUp: function() {
+			assert.ignore();
```
[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply tests/plugins/tableselection/integrations/core/selection.js]: That's some crazy stupid leftover by me. Thanks for spotting it, it would invalidate this test case on all browsers.

**Comment by jacekbogdanski** [comment]
It seems like embed based plugins somehow work on IE8, so I will wait with implementing custom `isSupportedEnvironment` member until we resolve this issue. Additionally, I extracted issue for easyimage base plugins https://github.com/ckeditor/ckeditor-dev/issues/3162 so we can do better research on this topic.

File: tests/plugins/cloudservices/cloudservices.js [hunk 10]
```diff
@@ -1,6 +1,5 @@
-/* bender-ckeditor-plugins: cloudservices */
+/* bender-ckeditor-plugins: cloudservices,easyimage */
 /* bender-include: %BASE_PATH%/plugins/easyimage/_helpers/tools.js*/
```
Comment (f1ames) [inline tests/plugins/cloudservices/cloudservices.js]: Is this line necessary? I see no usage of `easyImageTools `?

**Comment by jacekbogdanski** [comment]
Unfortunately, at this execution moment availability of `mathjax` plugin is not deterministic as the plugin is loaded asynchronously (e.g. it's available when console open, but most of the time it's not with normal execution), so we can't replace this condition easily.

**Comment by f1ames** [comment]
> Unfortunately, at this execution moment availability of mathjax plugin is not deterministic as the plugin is loaded asynchronously (e.g. it's available when console open, but most of the time it's not with normal execution), so we can't replace this condition easily.

I see, we would probably need to change how `ignoreStyle` works to be evaluated inside created test (so instead of simple Boolean would need to accept a function too). @jacekbogdanski do you think it would make sense or maybe you have any other idea in mind? This is a separate case so we could extract it to a separate issue, WDYT?

**Comment by f1ames** [comment]
Rebased on the latest `major`.

**Comment by jacekbogdanski** [comment]
Extracted issue: https://github.com/ckeditor/ckeditor-dev/issues/3173
